pkg/resources/peers.go

```diff
@@ -0,0 +1,167 @@
+package resources
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+	"github.com/jodevsa/wireguard-operator/pkg/api/v1alpha1"
+	"github.com/korylprince/ipnetgen"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Peers struct {
+	Wireguard *v1alpha1.Wireguard
+	Logger    logr.Logger
+	Client    client.Client
+}
+
+func (p Peers) Type() string {
+	return "peers"
+}
+
+func (p Peers) Converged(ctx context.Context) (bool, error) {
+	return true, nil
+}
+
+func (p Peers) Name() string {
+	return "peers"
+}
+
+func (p Peers) Exists(ctx context.Context) (bool, error) {
+	return true, nil
+}
+
+func (p Peers) Create(ctx context.Context) error {
+	return p.Update(ctx)
+}
+
+func (p Peers) NeedsUpdate(ctx context.Context) (bool, error) {
+	peers, err := p.getWireguardPeers(ctx)
+
+	for _, peer := range peers.Items {
+		if peer.Spec.Address == "" {
+			return true, nil
+		}
+
+		cfg := p.getPeerConfig(peer)
+		if peer.Status.Config != cfg || peer.Status.Status != v1alpha1.Ready {
+			return true, nil
+
+		}
+	}
+	if err != nil {
+		return true, nil
+	}
+
+	return false, nil
+}
+func (p Peers) getPeerConfig(peer v1alpha1.WireguardPeer) string {
+	cfg := fmt.Sprintf(`
```

Comment:
I wonder if we should actually encode real object instead of string formatting. We wouldn't do this for yaml or json, so why from toml?

Hypothetically:

```go
type WireguardConfiguration struct {
    Interface WireGuardInterfaceConfiguration
    // ...
}

type WireGuardInterfaceConfiguration struct {
    PrivateKey string
    Address string
    DNS string
}

b, err := toml.Marshal(WireguardConfiguration{
    Interface: WireGuardInterfaceConfiguration{
        PrivateKey: fmt.Sprintf("$(kubectl get secret %s-peer --template={{.data.privateKey}} -n %s | base64 -d)", peer.Name, peer.Namespace),
        Address: peer.Spec.Address
        DNS: p.constructDnsConfig()
    },
})
// ...
```